File: src/main/java/no/nav/sosialhjelp/soknad/business/db/SQLUtils.java
package no.nav.sosialhjelp.soknad.business.db;

import java.sql.Timestamp;
import java.time.LocalDateTime;

import static java.lang.System.getProperty;

public final class SQLUtils {

    private SQLUtils() {
    }

    private static final String HSQLDB = "hsqldb";
    public static final String DIALECT_PROPERTY = "sqldialect";

    public static String limit(int limit) {
        if (HSQLDB.equals(getProperty(DIALECT_PROPERTY))) {
            return "limit " + limit;
        } else {
            return "and rownum <= " + limit;
        }
    }

    public static String whereLimit(int limit) {
        if (HSQLDB.equals(getProperty(DIALECT_PROPERTY))) {
            return "limit " + limit;
        } else {
            return "where rownum <= " + limit;
        }
    }

    public static String toDate(int antallDager) {
        if (HSQLDB.equals(getProperty(DIALECT_PROPERTY))) {
            return "CURRENT_TIMESTAMP - " + antallDager + " DAY";
        } else {
            return "CURRENT_TIMESTAMP - NUMTODSINTERVAL("+antallDager+",'DAY') " ;
        }
    }

    public static String selectNextSequenceValue(String sequence) {
        if (HSQLDB.equals(getProperty(DIALECT_PROPERTY))) {
            return "call next value for " + sequence;
        } else {
            return "select " + sequence + ".nextval from dual";
        }
    }

    public static String selectMultipleNextSequenceValues(String sequence) {
        if (HSQLDB.equals(getProperty(DIALECT_PROPERTY))) {
            return "select next value for " + sequence + " from unnest(sequence_array(1,?,1))";
        } else {
            return "select " + sequence + ".nextval from dual connect by level <= ?";
        }
    }

    public static Timestamp tidTilTimestamp(LocalDateTime tid) {
        return tid != null
                ? Timestamp.valueOf(tid)
                : null;
    }

    public static LocalDateTime timestampTilTid(Timestamp timestamp) {
        return timestamp != null
                ? timestamp.toLocalDateTime()
                : null;
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/repository/SQLUtils.kt
package no.nav.sosialhjelp.soknad.repository

import java.sql.Timestamp
import java.time.LocalDateTime

object SQLUtils {

    private const val HSQLDB = "hsqldb"
    const val DIALECT_PROPERTY = "sqldialect"

    fun limit(limit: Int): String {
        return "limit $limit"
//
//        return if (HSQLDB == System.getProperty(DIALECT_PROPERTY)) {
//            "limit $limit"
//        } else {
//            "and rownum <= $limit"
//        }
    }

    fun selectNextSequenceValue(sequence: String): String {
        return if (HSQLDB == System.getProperty(DIALECT_PROPERTY)) {
            "call next value for $sequence"
        } else {
            "select $sequence.nextval from dual"
        }
    }

    fun tidTilTimestamp(tid: LocalDateTime?): Timestamp? {
        return tid?.let { Timestamp.valueOf(it) }
    }

    fun nullableTimestampTilTid(timestamp: Timestamp?): LocalDateTime? {
        return timestamp?.toLocalDateTime()
    }

    fun timestampTilTid(timestamp: Timestamp): LocalDateTime {
        return timestamp.toLocalDateTime()
    }

    fun blobSizeQuery(): String {
        return if (HSQLDB == System.getProperty(DIALECT_PROPERTY)) {
            "OCTET_LENGTH(DATA)"
        } else {
            "dbms_lob.getLength(DATA)"
        }
    }
}
